src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,171 @@
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element :list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return node(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> swapNode = node(index);
+        T oldVal = swapNode.item;
+        swapNode.item = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unLink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> currentNode = head; currentNode != null; currentNode = currentNode.next) {
+                if (currentNode.item == null) {
+                    unLink(currentNode);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> currentNode = head; currentNode != null; currentNode = currentNode.next) {
+                if (object.equals(currentNode.item)) {
+                    unLink(currentNode);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkPositionIndex(int index) {
+        if (!isPositionIndex(index)) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private boolean isPositionIndex(int index) {
+        return index >= 0 && index <= size;
+    }
+
+    private void checkElementIndex(int index) {
+        if (!isElementIndex(index)) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private boolean isElementIndex(int index) {
+        return index >= 0 && index < size;
+    }
+
+    void linkBefore(T element, Node<T> currentNode) {
+        final Node<T> before = currentNode.prev;
+        final Node<T> newNode = new Node<>(before, element, currentNode);
+        currentNode.prev = newNode;
+        if (before == null) {
+            head = newNode;
+        } else {
+            before.next = newNode;
+        }
+        size++;
+    }
+
+    Node<T> node(int index) {
+        Node<T> currentNode;
+        if (index < (size >> 1)) {
+            currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+        }
+        return currentNode;
+    }
+
+    T unLink(Node<T> currentNode) {
```

Comment:
```suggestion
    private T unLink(Node<T> currentNode) {
```
make helper methods private (also don't forget to check if your solution contains common mistakes)